actionview/lib/action_view/rendering.rb

```diff
@@ -35,24 +35,36 @@ def process(*) #:nodoc:
     end
 
     module ClassMethods
-      def view_context_class
-        @view_context_class ||= begin
-          supports_path = supports_path?
-          routes  = respond_to?(:_routes)  && _routes
-          helpers = respond_to?(:_helpers) && _helpers
-
-          Class.new(ActionView::Base) do
-            if routes
-              include routes.url_helpers(supports_path)
-              include routes.mounted_helpers
-            end
-
-            if helpers
-              include helpers
-            end
+      def _routes
+      end
+
+      def _helpers
+      end
+
+      def build_av_class(klass, supports_path, routes, helpers)
+        Class.new(klass) do
+          if routes
+            include routes.url_helpers(supports_path)
+            include routes.mounted_helpers
+          end
+
+          if helpers
+            include helpers
           end
         end
       end
+
+      def view_context_class
+        klass = ActionView::DetailsKey.view_context_class(ActionView::Base)
+
+        @view_context_class ||= build_av_class(klass, supports_path?, _routes, _helpers)
+
+        if klass.changed?(@view_context_class)
+          @view_context_class = build_av_class(klass, supports_path?, _routes, _helpers)
+        end
```

Comment:
Are we checking for changes from the DetailsKey view context class because we used to be clearing `@view_context_class` entirely?

Reply:
That's effectively what was happening.  [This middleware](https://github.com/rails/rails/blob/52af51c08a48ea7c6e0df09e2470a66a3141087d/actionview/lib/action_view/digestor.rb#L11-L15) would cause the [cache in the resolver](https://github.com/rails/rails/blob/52af51c08a48ea7c6e0df09e2470a66a3141087d/actionview/lib/action_view/template/resolver.rb#L73) to always miss.  Meaning the templates get recompiled every request.  This maintains the same behavior, but allows us to eliminate the finalizer since the methods are thrown away.